iroh-net/src/relay/codec.rs

```diff
@@ -264,8 +264,67 @@ impl Frame {
         }
     }
 
+    fn from_ws_vec(vec: Vec<u8>) -> anyhow::Result<Self> {
+        if vec.is_empty() {
+            bail!("error parsing relay::codec::Frame: too few bytes (0)");
+        }
+        let bytes = Bytes::from(vec);
+        let typ = FrameType::from(bytes[0]);
+        let frame = Self::from_bytes(typ, bytes.slice(1..))?;
+        Ok(frame)
+    }
+
+    fn into_ws_vec(self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        bytes.put_u8(self.typ().into());
+        self.write_to(&mut bytes);
+        bytes
+    }
+
+    /// Serializes this to bytes and wraps it in a websocket binary message.
+    pub(crate) fn into_ws_message(self) -> std::io::Result<tungstenite::Message> {
+        Ok(tungstenite::Message::binary(self.into_ws_vec()))
+    }
+
+    /// Serializes this to bytes and wraps it in a websocket binary message for use in wasm.
+    pub(crate) fn into_wasm_ws_message(self) -> std::io::Result<tokio_tungstenite_wasm::Message> {
+        Ok(tokio_tungstenite_wasm::Message::binary(self.into_ws_vec()))
+    }
+
+    /// Unwraps any binary messages received via websockets and parses them into `Frame`s.
+    ///
+    /// Ignores any non-binary websocket messages with a warning.
+    pub(crate) fn from_ws_message(
+        msg: tungstenite::Result<tungstenite::Message>,
+    ) -> Option<anyhow::Result<Self>> {
+        match msg {
+            Ok(tungstenite::Message::Binary(vec)) => Some(Self::from_ws_vec(vec)),
+            Ok(msg) => {
+                tracing::warn!(?msg, "Got websocket message of unsupported type, skipping.");
+                None
+            }
+            Err(e) => Some(Err(e.into())),
+        }
+    }
+
+    /// Unwraps any binary messages received via websockets (in wasm) and parses them into `Frame`s.
+    ///
+    /// Ignores any non-binary websocket messages with a warning.
+    pub(crate) fn from_wasm_ws_message(
+        msg: tokio_tungstenite_wasm::Result<tokio_tungstenite_wasm::Message>,
+    ) -> Option<anyhow::Result<Self>> {
```

Comment:
Why are these four not `TryFrom` impls?

Reply:
I'm used to just not do `*From` traits except for errors.
I agree this is one of the cases where using `TryFrom` really is appropriate :+1:

Reply:
Looking at this again - the `from_ws_message` and `from_wasm_ws_message` functions actually return an `Option<Result<...>>`, so you can't implement `TryFrom` for them. We *could* write `From` functions, but IMO that's now getting really far away from how `From` and `TryFrom` are used usually.

Perhaps it makes sense to bikeshed the function names?

Reply:
given they are fallible, maybe call them `try_from_ws_message`

Reply:
Meh - ended up moving these helpers out of `Frame` and just inline their logic into the `Stream` and `Sink` implementations.
That also uncovered some weird, unnecessary `?` from the `into_ws_message` call-site. I think the code just belongs closer to the `Stream` implementation, the logic was quite intertwined anyways (effectively some nested matches, this is now flattened).

Instead, it's only two new functions on `Frame` remaining: `Frame::encode_for_ws_msg` and `Frame::decode_from_ws_msg`.

Feels way better IMHO.